File: Wire-iOS/Sources/Components/TokenField/TokenField.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "TokenField.h"
#import "TokenField+Internal.h"

#import "Wire-Swift.h"

@class TokenizedTextView;

static NSString* ZMLogTag ZM_UNUSED = @"UI";

CGFloat const accessoryButtonSize = 32.0f;



@implementation TokenField

#pragma mark - Init

- (instancetype)init
{
    self = [super init];
    if (self) {
        [self setup];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    return self;
}

#pragma mark - Setup

- (void)setup
{
    self.currentTokens = [@[] mutableCopy];
    self.numberOfLines = NSUIntegerMax;
    
    [self setupDefaultAppearance];
    [self setupSubviews];
    [self setupConstraints];
    [self setupStyle];
}

- (void)setupDefaultAppearance
{
    [self setupFonts];
    _textColor = [UIColor blackColor];
    _lineSpacing = 8.0f;
    _hasAccessoryButton = NO;
    _tokenTitleVerticalAdjustment = 1;

    self.tokenTitleColor = [UIColor whiteColor];
    self.tokenSelectedTitleColor = [UIColor colorWithRed:0.103 green:0.382 blue:0.691 alpha:1.000];
    self.tokenBackgroundColor = [UIColor colorWithRed:0.118 green:0.467 blue:0.745 alpha:1.000];
    self.tokenSelectedBackgroundColor = [UIColor whiteColor];
    self.tokenBorderColor = [UIColor colorWithRed:0.118 green:0.467 blue:0.745 alpha:1.000];
    self.tokenSelectedBorderColor = [UIColor colorWithRed:0.118 green:0.467 blue:0.745 alpha:1.000];
    self.tokenTextTransform = TextTransformUpper;
    self.dotColor = [ColorScheme.defaultColorScheme colorWithName:ColorSchemeColorTextDimmed];
}

- (void)setupConstraints
{
    NSDictionary *views = @{@"textView": self.textView,
                            @"toLabel": self.toLabel,
                            @"button": self.accessoryButton};
    NSDictionary *metrics = @{@"left": @(self.textView.textContainerInset.left),
                              @"top": @(self.textView.textContainerInset.top),
                              @"right": @(self.textView.textContainerInset.right),
                              @"bSize": @(accessoryButtonSize),
                              @"bTop": @(self.accessoryButtonTop),
                              @"bRight": @(self.accessoryButtonRight),
                              };

    [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|[textView]|" options:0 metrics:nil views:views]];
    [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|[textView]|" options:0 metrics:nil views:views]];
    self.accessoryButtonRightMargin = [[NSLayoutConstraint constraintsWithVisualFormat:@"H:[button]-(bRight)-|" options:0 metrics:metrics views:views] objectAtIndex:0];
    self.accessoryButtonTopMargin = [[NSLayoutConstraint constraintsWithVisualFormat:@"V:|-(bTop)-[button]" options:0 metrics:metrics views:views] objectAtIndex:0];
    [self addConstraints:@[self.accessoryButtonRightMargin, self.accessoryButtonTopMargin]];
    [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:[button(bSize)]" options:0 metrics:metrics views:views]];
    [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:[button(bSize)]" options:0 metrics:metrics views:views]];
    
    self.toLabelLeftMargin = [[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-(left)-[toLabel]" options:0 metrics:metrics views:views] objectAtIndex:0];
    self.toLabelTopMargin = [[NSLayoutConstraint constraintsWithVisualFormat:@"V:|-(top)-[toLabel]" options:0 metrics:metrics views:views] objectAtIndex:0];
    [self.textView addConstraints:@[self.toLabelLeftMargin, self.toLabelTopMargin]];
    
    [self updateTextAttributes];
}

#pragma mark - Appearance

- (void)setFont:(UIFont *)font
{
    if ([_font isEqual:font]) {
        return;
    }
    _font = font;
    [self updateTextAttributes];
}

- (void)setTextColor:(UIColor *)textColor
{
    if ([_textColor isEqual:textColor]) {
        return;
    }
    _textColor = textColor;
    [self updateTextAttributes];
}

- (void)setLineSpacing:(CGFloat)lineSpacing
{
    if (_lineSpacing == lineSpacing) {
        return;
    }
    _lineSpacing = lineSpacing;
    [self updateTextAttributes];
}

- (void)setTokenOffset:(CGFloat)tokenOffset
{
    if (_tokenOffset == tokenOffset) {
        return;
    }
    _tokenOffset = tokenOffset;
    [self updateExcludePath];
    [self updateTokenAttachments];
}

- (NSDictionary *)textAttributes
{
    NSMutableDictionary *attributes = [NSMutableDictionary new];
    
    NSMutableParagraphStyle *inputParagraphStyle = [[NSMutableParagraphStyle alloc] init];
    inputParagraphStyle.lineSpacing = self.lineSpacing;
    attributes[NSParagraphStyleAttributeName] = inputParagraphStyle;
    
    if (self.font) {
        attributes[NSFontAttributeName] = self.font;
    }
    if (self.textColor) {
        attributes[NSForegroundColorAttributeName] = self.textColor;
    }
    
    return attributes;
}

- (void)setToLabelText:(NSString *)toLabelText
{
    if ([_toLabelText isEqualToString:toLabelText]) {
        return;
    }
    _toLabelText = toLabelText;
    [self updateTextAttributes];
}

- (void)setHasAccessoryButton:(BOOL)hasAccessoryButton
{
    if (_hasAccessoryButton == hasAccessoryButton) {
        return;
    }
    
    _hasAccessoryButton = hasAccessoryButton;
    self.accessoryButton.hidden = ! hasAccessoryButton;
    [self updateExcludePath];
}

- (void)setTokenTitleColor:(UIColor *)color
{
    if ([_tokenTitleColor isEqual:color]) {
        return;
    }
    _tokenTitleColor = color;
    [self updateTokenAttachments];
}

- (void)setTokenSelectedTitleColor:(UIColor *)color
{
    if ([_tokenSelectedTitleColor isEqual:color]) {
        return;
    }
    _tokenSelectedTitleColor = color;
    [self updateTokenAttachments];
}

- (void)setTokenBackgroundColor:(UIColor *)color
{
    if ([_tokenBackgroundColor isEqual:color]) {
        return;
    }
    _tokenBackgroundColor = color;
    [self updateTokenAttachments];
}

- (void)setTokenSelectedBackgroundColor:(UIColor *)color
{
    if ([_tokenSelectedBackgroundColor isEqual:color]) {
        return;
    }
    _tokenSelectedBackgroundColor = color;
    [self updateTokenAttachments];
}

- (void)setTokenBorderColor:(UIColor *)color
{
    if ([_tokenBorderColor isEqual:color]) {
        return;
    }
    _tokenBorderColor = color;
    [self updateTokenAttachments];
}

- (void)setTokenSelectedBorderColor:(UIColor *)color
{
    if ([_tokenSelectedBorderColor isEqual:color]) {
        return;
    }
    _tokenSelectedBorderColor = color;
    [self updateTokenAttachments];
}

- (void)setTokenTitleVerticalAdjustment:(CGFloat)tokenTitleVerticalAdjustment
{
    if (_tokenTitleVerticalAdjustment == tokenTitleVerticalAdjustment) {
        return;
    }
    _tokenTitleVerticalAdjustment = tokenTitleVerticalAdjustment;
    [self updateTokenAttachments];
}

#pragma mark - UIView overrides

- (BOOL)isFirstResponder
{
    return self.textView.isFirstResponder;
}

- (BOOL)canBecomeFirstResponder
{
    return [self.textView canBecomeFirstResponder];
}

- (BOOL)canResignFirstResponder
{
    return [self.textView canResignFirstResponder];
}

- (BOOL)becomeFirstResponder
{
    [self setCollapsed:NO animated:YES];
    return [self.textView becomeFirstResponder];
}

- (BOOL)resignFirstResponder
{
    [super resignFirstResponder];
    return [self.textView resignFirstResponder];
}

#pragma mark - Public Interface

- (NSArray *)tokens
{
    return self.currentTokens;
}

- (void)addTokenForTitle:(NSString *)title representedObject:(id)object
{
    Token *token = [[Token alloc] initWithTitle:title representedObject:object];
    [self addToken:token];
}

- (void)addToken:(Token *)token
{
    if (! [self.currentTokens containsObject:token]) {
        [self.currentTokens addObject:token];
    } else {
        return;
    }
    
    [self updateMaxTitleWidthForToken:token];
    
    if (! self.isCollapsed) {
        self.textView.attributedText = [self stringForTokens:self.currentTokens];
        // Calling -insertText: forces textView to update its contentSize, while other public methods do not.
        // Broken contentSize leads to broken scrolling to bottom of input field.
        [self.textView insertText:@""];
        
        if ([self.delegate respondsToSelector:@selector(tokenField:changedFilterTextTo:)]) {
            [self.delegate tokenField:self changedFilterTextTo:@""];
        }

        [self invalidateIntrinsicContentSize];
        
        // Move the cursor to the end of the input field
        self.textView.selectedRange = NSMakeRange(self.textView.text.length, 0);
        
        // autoscroll to the end of the input field
        [self setNeedsLayout];
        [self updateLayout];
        [self scrollToBottomOfInputField];
    } else {
        self.textView.attributedText = [self collapsedString];
        [self invalidateIntrinsicContentSize];
    }    
}

- (void)updateMaxTitleWidthForToken:(Token *)token
{
    CGFloat tokenMaxSizeWidth = self.textView.textContainer.size.width;
    if (self.currentTokens.count == 0) {
        tokenMaxSizeWidth -= self.toLabel.frame.size.width + (self.hasAccessoryButton ? self.accessoryButton.frame.size.width : 0.0f) + self.tokenOffset;
    } else if (self.currentTokens.count == 1) {
        tokenMaxSizeWidth -= (self.hasAccessoryButton ? self.accessoryButton.frame.size.width : 0.0f);
    }
    token.maxTitleWidth = tokenMaxSizeWidth;
}

- (Token *)tokenForRepresentedObject:(id)object
{
    for (Token *token in self.currentTokens) {
        if ([token.representedObject isEqual:object]) {
            return token;
        }
    }
    return nil;
}

- (void)scrollToBottomOfInputField
{
    if (self.textView.contentSize.height > self.textView.bounds.size.height) {
        [self.textView setContentOffset:CGPointMake(0.0f, self.textView.contentSize.height - self.textView.bounds.size.height)
                               animated:YES];
    }
    else {
        [self.textView setContentOffset:CGPointZero];
    }
}

- (void)setExcludedRect:(CGRect)excludedRect
{
    if (CGRectEqualToRect(_excludedRect, excludedRect)) {
        return;
    }
    
    _excludedRect = excludedRect;
    [self updateExcludePath];
}

- (void)setNumberOfLines:(NSUInteger)numberOfLines
{
    if (_numberOfLines != numberOfLines) {
        _numberOfLines = numberOfLines;
        [self invalidateIntrinsicContentSize];
    }
}

- (void)setCollapsed:(BOOL)collapsed
{
    [self setCollapsed:collapsed animated:NO];
}

- (void)setCollapsed:(BOOL)collapsed animated:(BOOL)animated
{
    if (_collapsed == collapsed) {
        return;
    }
    
    if (self.currentTokens.count == 0) {
        return;
    }
    
    _collapsed = collapsed;
    
    dispatch_block_t animationBlock = ^{
        [self invalidateIntrinsicContentSize];
        [self layoutIfNeeded];
    };
    ZM_WEAK(self);
    void (^compeltionBlock) (BOOL) = ^(BOOL finnished) {
        ZM_STRONG(self);
        if (self.collapsed) {
            self.textView.attributedText = [self collapsedString];
            [self invalidateIntrinsicContentSize];
            [UIView animateWithDuration:0.2 animations:^{
                [self.textView setContentOffset:CGPointZero animated:NO];
            }];
        } else {
            self.textView.attributedText = [self stringForTokens:self.currentTokens];
            [self invalidateIntrinsicContentSize];
            if (self.textView.attributedText.length > 0) {
                self.textView.selectedRange = NSMakeRange(self.textView.attributedText.length, 0);
                [UIView animateWithDuration:0.2 animations:^{
                    [self.textView scrollRangeToVisible:self.textView.selectedRange];
                }];
            }
        }
    };
    
    if (animated) {
        [UIView animateWithDuration:0.25
                         animations:animationBlock
                         completion:compeltionBlock];
    } else {
        animationBlock();
        compeltionBlock(YES);
    }
}

#pragma mark - Layout

- (CGSize)intrinsicContentSize
{
    CGFloat height = self.textView.contentSize.height;
    CGFloat maxHeight = self.font.lineHeight * self.numberOfLines + self.lineSpacing * (self.numberOfLines - 1) +
        self.textView.textContainerInset.top + self.textView.textContainerInset.bottom;
    CGFloat minHeight = self.font.lineHeight * 1 + self.textView.textContainerInset.top + self.textView.textContainerInset.bottom;
    
    if (self.collapsed) {
        return CGSizeMake(UIViewNoIntrinsicMetric, minHeight);
    } else {
        return CGSizeMake(UIViewNoIntrinsicMetric, MAX(MIN(height, maxHeight), minHeight));
    }
}

- (CGFloat)accessoryButtonTop
{
    return self.textView.textContainerInset.top + (self.font.lineHeight - accessoryButtonSize) / 2 - self.textView.contentOffset.y;
}

- (CGFloat)accessoryButtonRight
{
    return self.textView.textContainerInset.right;
}

- (void)updateLayout
{
    if (self.toLabelText.length > 0) {
        self.toLabelLeftMargin.constant = self.textView.textContainerInset.left;
        self.toLabelTopMargin.constant = self.textView.textContainerInset.top;
    }
    if (self.hasAccessoryButton) {
        self.accessoryButtonRightMargin.constant = self.accessoryButtonRight;
        self.accessoryButtonTopMargin.constant = self.accessoryButtonTop;
    }
    [self layoutIfNeeded];
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    
    BOOL anyTokenUpdated = NO;
    for (Token *token in self.currentTokens) {
        if (token.maxTitleWidth == 0) {
            [self updateMaxTitleWidthForToken:token];
            anyTokenUpdated = YES;
        }
    }
    
    if (anyTokenUpdated) {
        [self updateTokenAttachments];
        NSRange wholeRange = NSMakeRange(0, self.textView.attributedText.length);
        [self.textView.layoutManager invalidateLayoutForCharacterRange:wholeRange actualCharacterRange:NULL];
    }
}

#pragma mark - Utility

- (NSAttributedString *)collapsedString
{
    NSString *collapsedText = NSLocalizedString(@" ...", nil);

    return [[NSAttributedString alloc] initWithString:collapsedText attributes:self.textAttributes];
}

- (void)clearFilterText
{
    __block NSInteger firstCharacterIndex = NSNotFound;
    
    NSCharacterSet *notWhitespace = [[NSCharacterSet whitespaceAndNewlineCharacterSet] invertedSet];
    
    [self.textView.text enumerateSubstringsInRange:NSMakeRange(0, self.textView.text.length)
                                           options:NSStringEnumerationByComposedCharacterSequences
                                        usingBlock:^(NSString *substring, NSRange substringRange, NSRange enclosingRange, BOOL *stop)
     {
         if (substring.length && ([substring characterAtIndex:0] != NSAttachmentCharacter) && [substring rangeOfCharacterFromSet:notWhitespace].location != NSNotFound) {
             firstCharacterIndex = substringRange.location;
             *stop = YES;
         }
     }];
    
    self.filterText = @"";
    if (firstCharacterIndex != NSNotFound) {
        NSRange rangeToClear = NSMakeRange(firstCharacterIndex, self.textView.text.length - firstCharacterIndex);
        
        [self.textView.textStorage beginEditing];
        [self.textView.textStorage deleteCharactersInRange:rangeToClear];
        [self.textView.textStorage endEditing];
        [self.textView insertText:@""];
        
        [self invalidateIntrinsicContentSize];
        [self layoutIfNeeded];
    }
}

- (void)updateTextAttributes
{
    self.textView.typingAttributes = self.textAttributes;
    [self.textView.textStorage beginEditing];
    [self.textView.textStorage addAttributes:self.textAttributes range:NSMakeRange(0, self.textView.textStorage.length)];
    [self.textView.textStorage endEditing];
    
    if (self.toLabelText) {
        self.toLabel.attributedText = [[NSMutableAttributedString alloc] initWithString:self.toLabelText attributes:self.textAttributes];
    } else {
        self.toLabel.text = @"";
    }
    
    [self updateExcludePath];
}

- (void)updateExcludePath
{
    [self updateLayout];
    
    NSMutableArray *exclusionPaths = [@[] mutableCopy];
    
    if (CGRectEqualToRect(self.excludedRect, CGRectZero) == false) {
        CGAffineTransform transform = CGAffineTransformMakeTranslation(self.textView.contentOffset.x,
                                                                       self.textView.contentOffset.y);
        CGRect transformedRect = CGRectApplyAffineTransform(self.excludedRect, transform);
        UIBezierPath *path = [UIBezierPath bezierPathWithRect:transformedRect];
        [exclusionPaths addObject:path];
    }
    
    if (self.toLabelText.length > 0) {
        CGRect transformedRect = CGRectOffset(self.toLabel.frame,
                                              - self.textView.textContainerInset.left,
                                              - self.textView.textContainerInset.top);
        transformedRect.size.width += self.tokenOffset;
        UIBezierPath *path = [UIBezierPath bezierPathWithRect:transformedRect];
        [exclusionPaths addObject:path];
    }
    
    if (self.hasAccessoryButton) {
        // Exclude path should be relative to content of button, not frame.
        // Assuming intrinsic content size is a size of visual content of the button,
        // 1. Calcutale frame with same center as accessoryButton has, but with size of intrinsicContentSize
        CGRect transformedRect = self.accessoryButton.frame;
        CGSize contentSize = CGSizeMake(accessoryButtonSize, accessoryButtonSize);
        transformedRect = CGRectInset(transformedRect, 0.5 * (transformedRect.size.width - contentSize.width), 0.5 * (transformedRect.size.height - contentSize.height));
        
        // 2. Convert frame to textView coordinate system
        transformedRect = [self.textView convertRect:transformedRect fromView:self];
        CGAffineTransform transform = CGAffineTransformMakeTranslation( - self.textView.textContainerInset.left, - self.textView.textContainerInset.top);
        transformedRect = CGRectApplyAffineTransform(transformedRect, transform);
        
        UIBezierPath *path = [UIBezierPath bezierPathWithRect:transformedRect];
        [exclusionPaths addObject:path];
    }
    
    self.textView.textContainer.exclusionPaths = exclusionPaths;
}


#pragma mark - UIScrollViewDelegate

- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    if (scrollView == self.textView) {
        [self updateExcludePath];
    }
}

@end

File: Wire-iOS/Sources/Components/TokenField/TokenField.swift
//
// Wire
// Copyright (C) 2018 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import UIKit

private let zmLog = ZMSLog(tag: "TokenField")

extension TokenField {

    @objc
    func setupSubviews() {
        // this prevents accessoryButton to be visible sometimes on scrolling
        clipsToBounds = true

        let textView = TokenizedTextView()
        textView.tokenizedTextViewDelegate = self
        textView.delegate = self
        textView.translatesAutoresizingMaskIntoConstraints = false
        textView.backgroundColor = UIColor.clear
        if #available(iOS 11, *) {
            textView.textDragInteraction?.isEnabled = false
        }
        addSubview(textView)

        toLabel = UILabel()
        toLabel.translatesAutoresizingMaskIntoConstraints = false
        toLabel.font = font
        toLabel.text = toLabelText
        toLabel.backgroundColor = UIColor.clear
        textView.addSubview(toLabel)

        self.textView = textView

        // Accessory button could be a subview of textView,
        // but there are bugs with setting constraints from subview to UITextView trailing.
        // So we add button as subview of self, and update its position on scrolling.
        accessoryButton = IconButton()
        accessoryButton.translatesAutoresizingMaskIntoConstraints = false
        accessoryButton.isHidden = !hasAccessoryButton
        addSubview(accessoryButton)
    }

    @objc func setupStyle() {
        tokenOffset = 4

        textView.tintColor = .accent()
        textView.autocorrectionType = .no
        textView.returnKeyType = .go
        textView.placeholderFont = .smallRegularFont
        textView.placeholderTextContainerInset = UIEdgeInsets(top: 0, left: 48, bottom: 0, right: 0)
        textView.placeholderTextTransform = .upper
        textView.lineFragmentPadding = 0
    }

    @objc func setupFonts() {
        // Dynamic Type is disabled for now until the separator dots
        // vertical alignment has been fixed for larger fonts.
        let schema = FontScheme(contentSizeCategory: .medium)
        font = schema.font(for: .init(.normal, .regular))
        tokenTitleFont = schema.font(for: .init(.small, .regular))
    }

    // MARK: - Utility

    @objc
    func updateTokenAttachments() {
        textView?.attributedText.enumerateAttachment() { tokenAttachment, _, _ in
            (tokenAttachment as? TokenTextAttachment)?.refreshImage()
        }
    }

    @objc
    func string(forTokens tokens: [Token]) -> NSAttributedString {
        let string = NSMutableAttributedString()
        for token in tokens {
            let tokenAttachment = TokenTextAttachment(token: token, tokenField: self)
            let tokenString = NSAttributedString(attachment: tokenAttachment)

            string.append(tokenString)

            let separatorAttachment = TokenSeparatorAttachment(token: token, tokenField: self)
            let separatorString = NSAttributedString(attachment: separatorAttachment)

            string.append(separatorString)
        }

        return string && (textAttributes as? [NSAttributedString.Key: Any]) ?? [:]
    }

    /// update currentTokens with textView's current attributedText text after the textView change the text
    func filterUnwantedAttachments() {
        var updatedCurrentTokens: Set<Token> = []
        var updatedCurrentSeparatorTokens: Set<Token> = []

        textView.attributedText.enumerateAttachment() { textAttachment, _, _ in

            if let token = (textAttachment as? TokenTextAttachment)?.token,
                !updatedCurrentTokens.contains(token) {
                updatedCurrentTokens.insert(token)
            }

            if let token = (textAttachment as? TokenSeparatorAttachment)?.token,
                !updatedCurrentSeparatorTokens.contains(token) {
                updatedCurrentSeparatorTokens.insert(token)
            }
        }

        updatedCurrentTokens = updatedCurrentTokens.intersection(updatedCurrentSeparatorTokens)

        ///TODO: Change currentTokens type to [Token]
        if let currentTokens = self.currentTokens as? [Token] {
            var deletedTokens = Set<Token>(currentTokens)
            deletedTokens.subtract(updatedCurrentTokens)

            if !deletedTokens.isEmpty {
                removeTokens(Array(deletedTokens))
            }

            self.currentTokens.removeObjects(in: Array(deletedTokens))
            delegate?.tokenField(self, changedTokensTo: currentTokens)
        }
    }

    // MARK: - remove token

    func removeAllTokens() {
        removeTokens(currentTokens as! [Token])
        textView.showOrHidePlaceholder()
    }

    func removeToken(_ token: Token) {
        removeTokens([token])
    }

    private func removeTokens(_ tokensToRemove: [Token]) {
        var rangesToRemove: [NSRange] = []

        textView.attributedText.enumerateAttachment() { textAttachment, range, _ in
            if let token = (textAttachment as? TokenContainer)?.token,
                tokensToRemove.contains(token) {
                rangesToRemove.append(range)
            }
        }

        // Delete ranges from the end of string till the beginning: this keeps range locations valid.
        rangesToRemove.sort(by: { rangeValue1, rangeValue2 in
            rangeValue1.location > rangeValue2.location
        })

        textView.textStorage.beginEditing()
        for rangeValue in rangesToRemove {
            textView.textStorage.deleteCharacters(in: rangeValue)
        }
        textView.textStorage.endEditing()

        currentTokens?.removeObjects(in: tokensToRemove)

        invalidateIntrinsicContentSize()
        updateTextAttributes()

        textView.showOrHidePlaceholder()
    }

    private func rangeIncludesRange(_ range: NSRange, _ includedRange: NSRange) -> Bool {
        return range == range.union(includedRange)
    }

    private func notifyIfFilterTextChanged() {
        var indexOfFilterText = 0
        textView.attributedText.enumerateAttachment() { tokenAttachment, range, _ in
            if tokenAttachment is TokenTextAttachment {
                indexOfFilterText = range.upperBound
            }
        }

        let oldFilterText = filterText
        self.filterText = ((textView.text as NSString).substring(from: indexOfFilterText)).replacingOccurrences(of: "\u{FFFC}", with: "")
        if oldFilterText != filterText {
            delegate?.tokenField(self, changedFilterTextTo: filterText)
        }
    }

}

// MARK: - TokenizedTextViewDelegate

extension TokenField: TokenizedTextViewDelegate {
    func tokenizedTextView(_ textView: TokenizedTextView, didTapTextRange range: NSRange, fraction: CGFloat) {
        if isCollapsed {
            setCollapsed(false, animated: true)
            return
        }

        if fraction >= 1 && range.location == self.textView.textStorage.length - 1 {
            return
        }

        if range.location < textView.textStorage.length {
            textView.attributedText.enumerateAttachment() { tokenAttachemnt, range, _ in
                if tokenAttachemnt is TokenTextAttachment {
                    textView.selectedRange = range
                }
            }
        }
    }

    func tokenizedTextView(_ textView: TokenizedTextView, textContainerInsetChanged textContainerInset: UIEdgeInsets) {
        invalidateIntrinsicContentSize()
        updateExcludePath()
        updateLayout()
    }

}

// MARK: - UITextViewDelegate

extension TokenField: UITextViewDelegate {
    public func textView(_ textView: UITextView, shouldInteractWith textAttachment: NSTextAttachment, in characterRange: NSRange, interaction: UITextItemInteraction) -> Bool {
        return !(textAttachment is TokenSeparatorAttachment)
    }

    public func textViewDidChange(_ textView: UITextView) {
        userDidConfirmInput = false

        filterUnwantedAttachments()
        notifyIfFilterTextChanged()
        invalidateIntrinsicContentSize()
    }

    public func textViewDidChangeSelection(_ textView: UITextView) {
        zmLog.debug("Selection changed: NSStringFromRange(textView.selectedRange)")

        var modifiedSelectionRange = NSRange(location: 0, length: 0)
        var hasModifiedSelection = false

        textView.attributedText.enumerateAttachment() { tokenAttachment, range, _ in
            if let tokenAttachment = tokenAttachment as? TokenTextAttachment {
                tokenAttachment.isSelected = rangeIncludesRange(textView.selectedRange, range)
                textView.layoutManager.invalidateDisplay(forCharacterRange: range)

                if rangeIncludesRange(textView.selectedRange, range) {
                    modifiedSelectionRange = (hasModifiedSelection ? modifiedSelectionRange : range).union(range)
                    hasModifiedSelection = true
                }
                zmLog.info("    person attachement: \(tokenAttachment.token.title) at range: \(range) selected: \(tokenAttachment.isSelected)")
            }
        }

        if hasModifiedSelection, textView.selectedRange != modifiedSelectionRange {
            textView.selectedRange = modifiedSelectionRange
        }
    }

    public func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
        if (text == "\n") {
            textView.resignFirstResponder()
            userDidConfirmInput = true
            delegate?.tokenFieldDidConfirmSelection(self)

            return false
        }

        if range.length == 1 && text.isEmpty {
            // backspace
            var cancelBackspace = false
            textView.attributedText.enumerateAttachment() { tokenAttachment, range, stop in
                if let tokenAttachment = tokenAttachment as? TokenTextAttachment {
                    if !tokenAttachment.isSelected {
                        textView.selectedRange = range
                        cancelBackspace = true
                    }
                    
                    stop.pointee = true
                }
            }
            
            if cancelBackspace {
                return false
            }
        }

        // Inserting text between tokens does not make sense for this control.
        // If there are any tokens after the insertion point, move the cursor to the end instead, but only for insertions
        // If the range length is >0, we are trying to replace something instead, and that’s a bit more complex,
        // so don’t do any magic in that case
        if !text.isEmpty {
            let range = NSRange(location: range.location, length: textView.text.count - range.location)
            (textView.text as NSString).enumerateSubstrings(in: range,
                                                            options: .byComposedCharacterSequences,
                                                            using: { substring, _, _, stop in

                if substring?.isEmpty == false,
                    let nsString: NSString = substring as NSString?,
                    nsString.character(at: 0) == NSTextAttachment.character {
                    textView.selectedRange = NSRange(location: textView.text.count, length: 0)
                    stop.pointee = true
                }
            })
        }

        updateTextAttributes()

        return true

    }

}

extension NSAttributedString {
    func enumerateAttachment(block: (Any?, NSRange, UnsafeMutablePointer<ObjCBool>) -> Void) {
        enumerateAttribute(.attachment, in: NSRange(location: 0, length: length), options: [], using: block)
    }
}
